rust/xetblob/src/xet_repo.rs

```diff
@@ -856,41 +876,65 @@ impl XetRepoWriteTransaction {
                 }
                 NewFileSource::NewFile(ref f) => {
                     if let Some(contents) = f.closed_state().await {
+                        total_size += contents.len();
                         let action = Action {
                             action: "upsert".to_string(),
-                            file_path: i.0.clone(),
+                            file_path: file_id.0.clone(),
                             previous_path: String::new(),
                             execute_filemode: false,
                             content: contents.iter().map(|b| *b as char).collect::<String>(),
                         };
+                        total_size += action.content.len();
+
                         actions.push(action);
+
+                        if total_size >= TARGET_SINGLE_COMMIT_MAX_SIZE {
+                            commit_boundaries.push(actions.len());
+                            total_size = 0;
+                        }
                     }
                 }
             }
         }
-        for i in self.move_files.iter() {
-            let action = Action {
-                action: "move".to_string(),
-                file_path: i.1.clone(),
-                previous_path: i.0.clone(),
-                execute_filemode: false,
-                content: String::new(),
+
+        commit_boundaries.push(actions.len());
+
+        let mut upload_tasks = JoinSet::<Result<(), anyhow::Error>>::new();
+
+        for (lb_i, ub_i) in commit_boundaries[..(commit_boundaries.len() - 1)]
+            .iter()
+            .zip(&commit_boundaries[1..])
+        {
+            if *lb_i == *ub_i {
+                continue;
+            }
+
+            let local_actions: Vec<_> = ((*lb_i)..(*ub_i)).map(|i| take(&mut actions[i])).collect();
+
+            let command = JSONCommand {
+                author_name: author_name.to_string(),
+                author_email: author_email.to_string(),
+                branch: self.branch.clone(),
+                commit_message: commit_message.to_string(),
+                actions: local_actions,
+                create_ref: false,
             };
-            actions.push(action);
+            let git_object_commit_command = serde_json::to_string(&command)
+                .map_err(|_| anyhow!("Unexpected serialization error for {:?}", command))?;
+            info!("{git_object_commit_command}");
+
+            let remote_base_url = self.remote_base_url.clone();
+
+            upload_tasks.spawn(async move {
+                perform_atomic_commit_query(remote_base_url, &git_object_commit_command).await?;
+                Ok(())
+            });
```

Comment:
I am not sure we want to send the atomic_commit concurrently as that will just generate a bunch of load on the server. I don't expect us to get too much perf benefit here since the server can only actually commit one-by-one, so having many parallel commits just means that some of them will be forced to wait for the lock (which, notably, is an unfair lock), risking timeouts. 

True, there is some parallelism on the server-side that can speed things up (JSON parsing, verification, and preparation of the commit), which would improve performance, but I think we should test out the impact of this parallelism in a separate investigation / PR.